File: LeetCode/cpp/LCA.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>

using namespace std;

struct ListNode{
    int val;
    ListNode *next;
    ListNode(int x) : val(0), next(nullptr) {}
    ListNode(int x, ListNode *node) : val(x), next(node) {}
};

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *_left, TreeNode *_right) : val(x), left(_right), right(_right) {}
};


// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(NULL) {}
//     ListNode(int x) : val(x), next(NULL) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };

// struct TreeNode {
//     int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode() : val(0), left(NULL), right(NULL) {}
//     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
//     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
// };


class Solution{
    public:
        TreeNode* lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q){
            if (root == nullptr)
                return nullptr;
            
            if (root == p || root == q)
                return root;

            TreeNode *left = lowestCommonAncestor(root->left, p, q);
            TreeNode *right = lowestCommonAncestor(root->right, p, q);

            if (left != nullptr && right != nullptr)
                return root;
                        
            return (left != nullptr) ? left : right;
        }
};


// // 递归寻找最低公共祖先
// class Solution {
// public:
   
//     TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
//         if(root == nullptr)
//             return nullptr;
//         if(root == p || root == q)
//             return root;
        
//         TreeNode* left = lowestCommonAncestor(root->left, p, q);
//         TreeNode* right = lowestCommonAncestor(root->right, p, q);
        
//         if(left != nullptr && right != nullptr)
//             return root;
//         return (left != nullptr) ? left : right;
//     }
// };

File: LeetCode/python/LCA.py
"""
寻找二叉树最低公共祖先
"""


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.lchild = None
        self.rchild = None


def backtrack(root, valA, valB):
    if root is None:
        return None
    if root.val == valA or root.val == valB:
        return root

    left = backtrack(root.lchild, valA, valB)
    right = backtrack(root.rchild, valA, valB)
    if left is not None and right is not None:
        return root
    elif right is not None:
        return right
    else:
        return left


def lca(root, valA, valB):
    return backtrack(root, valA, valB)


if __name__ == "__main__":
    bst_root = TreeNode(8)
    bst_root.lchild = TreeNode(0)
    bst_root.rchild = TreeNode(20)

    bst_root.lchild.lchild = TreeNode(-2)
    bst_root.lchild.rchild = TreeNode(4)
    bst_root.rchild.lchild = TreeNode(15)
    bst_root.rchild.rchild = TreeNode(30)

    bst_root.rchild.lchild.lchild = TreeNode(10)
    bst_root.rchild.lchild.rchild = TreeNode(16)
    bst_root.rchild.rchild.lchild = TreeNode(25)
    bst_root.rchild.rchild.rchild = TreeNode(35)

    node = lca(bst_root, 10, 35)
    print(node.val)
